Fixup DeployToUpgrade

penev92 opened the pull request:
This fixes issues with `DeployToUpgrade` that lead to crashes, as well as adding support for having 3 separate actor states - `Undeployed, Deploying, Deployed`, with the ability to have 3 different sets of traits enabled according to the state (needed for more complex transformations like Tick Tanks).

Fixes #9242.
Fixes #9996.

penev92 committed:
Add support for deploy animations using a specific upgrade

Also ignore orders while playing a deploy/undeploy animation. Fixes 9242.

We grant a level 1 Undeployed upgrade(s) when the actor is created which enables the default behavior of the actor. When the actor receives a deploy order, the Undeployed upgrade(s) is(are) revoked, which may or may not be used to enable a trait to play a deploy animation. Once the deployment is complete, the Deployed upgrade(s) is(are) granted.
Then do the reverse when undeploying, disabling weapons and other systems first by revoking the Deployed upgrade(s), and granting the Undeployed upgrade(s) when the undeploy animation is done.

penev92 committed:
Update thumpers and Mobile Sensor Arrays to the new DeployToUpgrade

Currently those are the only two usages of the trait.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -18,11 +17,18 @@
 	{
+		[UpgradeGrantedReference]
+		[Desc("An upgrade that is granted while the actor is not deployed" +
+			" or in the process of deploying/undeploying.")]
```
This desc needs updating for the new usage.

penev92 replied:
This desc *was* updated for the new usage?

penev92 replied:
"is not deployed or in the process of deploying"
But OK.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -18,11 +17,18 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
+	enum DeployState { Undeployed, Deploying, Deployed }
```
This looks to be used only internally, so it can move inside the class.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -73,6 +79,11 @@ public DeployToUpgrade(Actor self, DeployToUpgradeInfo info)
 
+		public void Created(Actor self)
+		{
+			GrantUndeployedUpgrades();
```
At some point we'll want a `DeployedInit`, but that can be left to a future PR.

abcdefg30 replied:
#9501

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -89,59 +100,25 @@ public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool qu
-					else
-						RevokeUpgrades();
-				}));
+				self.CancelActivity();
```
This will behave badly if the order is queued.  I think at a minimum this needs to be pulled above the if, and live in its own `if (!order.Queued)`.  @obrakmann may be able to find some other ways to break this based on our recent experience with the `Transforms` trait.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -181,15 +158,81 @@ bool IsOnValidRampType()
+		/// <summary>Play undeploy sound and animation and after that revoke the upgrades.</summary>
+		void Undeploy()
+		{
+			// Something went wrong, most likely due to deploy order spam and the fact that this is a delayed action.
```
There's a comment here, but not on the check above.  Can we have a comment on both or on neither?

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -181,15 +158,81 @@ bool IsOnValidRampType()
+		{
+			deployState = DeployState.Undeployed;
+
+			if (info.UndeployedUpgrades.Length == 0)
```
The `foreach` is happy to loop 0 times, so this check isn't needed.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -181,15 +158,81 @@ bool IsOnValidRampType()
-			foreach (var up in info.Upgrades)
+			deployState = DeployState.Deploying;
+
+			if (info.UndeployedUpgrades.Length == 0)
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -181,15 +158,81 @@ bool IsOnValidRampType()
+				GrantUndeployedUpgrades();
+		}
+
+		void GrantDeployedUpgrades()
```
These methods are doing more than just granting upgrades, so IMO it would be clearer to call them `DeployCompleted` etc based on when they are called.

penev92 commented:
Updated.

pchote commented:
Seems to work well except for those two queueing issues and the comment nit (commented on the right PR this time).

obrakmann commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -89,59 +100,27 @@ public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool qu
+			if (order.OrderString != "DeployToUpgrade" || deployState == DeployState.Deploying)
 				return;
 
 			if (!IsOnValidTerrain())
```
If the order is queued this check doesn't make sense. Check `order.TargetLocation` instead of `self.Location`.

obrakmann replied:
Eh, it turns out that neither `TargetLocation` nor `ExtraLocation` are set in the order (which kinda makes sense, considering that you can just (un)deploy using a hotkey). I propose this change instead:

```diff
diff --git a/OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs b/OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs
index 523e879..84a2b53 100644
--- a/OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs
+++ b/OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs
@@ -103,9 +103,6 @@ public void ResolveOrder(Actor self, Order order)
                        if (order.OrderString != "DeployToUpgrade" || deployState == DeployState.Deploying)
                                return;
 
-                       if (!IsOnValidTerrain())
-                               return;
-
                        if (!order.Queued)
                                self.CancelActivity();
 
@@ -167,6 +164,9 @@ void Deploy()
                        if (deployState != DeployState.Undeployed)
                                return;
 
+                       if (!IsOnValidTerrain())
+                               return;
+
                        if (!string.IsNullOrEmpty(info.DeploySound))
                                Game.Sound.Play(info.DeploySound, self.CenterPosition);
```

obrakmann commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -89,59 +100,27 @@ public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool qu
 
-			if (isUpgraded)
+			if (!order.Queued)
+				self.CancelActivity();
```
```
(22:40:26) antares79: pchote: I think the queueing not working properly on #10014 is because the Move activity is doing freaky shit again
(22:41:07) antares79: at the point where my breakpoint stops, there's not NextActivty on MoveSecondHalf, although there are three more waypoints the unit is supposed to go to
(22:48:44) antares79: pchote: confirmed: [see image]
(22:49:16) antares79: the turn and callfunc are from DTU, the two moves are me ordering the thumper around before deploying
```
![move-activity-madness](https://cloud.githubusercontent.com/assets/4331210/11195879/aa4230f6-8cb4-11e5-8059-6f001d3e712c.png)

penev92 commented:
Updated the commend and moved the terrain check.
Couldn't find anything wrong with either thumper or mobile sensor array with both queued and non-queued orders.

obrakmann commented:
Looks ok to me now. :+1:

abcdefg30 commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -20,9 +19,13 @@ namespace OpenRA.Mods.Common.Traits
-		[Desc("The upgrades to grant when deploying and revoke when undeploying.")]
-		public readonly string[] Upgrades = { };
+		[Desc("The upgrades to grant after deploying and revoke before undeploying.")]
+		public readonly string[] DeployedUpgrades = { };
```
Doesn't this need an upgrade rule?

pchote replied:
Have we shipped a release with this yet?

abcdefg30 replied:
https://github.com/OpenRA/OpenRA/commit/0717ca57ea4b915b97e38d87a73ce321fcb85a5c#diff-7a3a62531cce0e8fa71a8b6ea112fcc3R22
`release-20150919 … playtest-20150524`

pchote replied:
Needs an upgrade rule then, sorry :/

penev92 replied:
> [21:32] <+penev> abcdefg30: basically I'm not gonna do a half-assed upgrade rule (because it won't be able to do the job fully anyway) for the *one* use-case in Nolt's mod, which would even get merged into prep

abcdefg30 replied:
https://github.com/penev92/OpenRA/pull/9

penev92 replied:
Merged, thanks.

abcdefg30 committed:
Add an upgrade rule for the Upgrades -> DeployedUpgrades change

abcdefg30 commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -181,16 +156,80 @@ bool IsOnValidRampType()
+
+			// If there is no animation to play just grant the upgrades that are used while undeployed.
+			// Alternatively, play the undeploy animation and then grant the upgrades.
+			if (string.IsNullOrEmpty(info.DeployAnimation))
```
What about the `body.Value == null` check here you have in `Deploy()`?

abcdefg30 commented:
Looks good to me now. :+1: / :white_check_mark:

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/df882b3ca67a66f075732f3f974f16c418ec5223...009ae8acf8311220a254f3fcd6f6ffe4a495877e)
